(SIMP-MAINT) Enable sphinx-autobuild server

op-ct opened the pull request:
This patch updates the sphinx-autobuild support in `rake docs:server`:

  * Fixes a bug that disabled sphinx-autobuild
  * Adds sphinx-autobuild to `requirements.txt`
  * Adds a `run_streaming_command` method to stream output while `rake
    docs:server` is running
  * Adds `SIMP_DOC_extra_sphinx_args=` support
    (same behavior as the other `rake docs:*`  build tasks)
  * Adds environment variables to the task documentation
  * Updates sphinx-autobuild invocation to ignore vim swap files

----

In response to feedback, I tried to get rid of the  warning messages
that appear when `SIMP_FAST_DOCS=true`, but only with partial success.
Sphinx is presently incapable of conditional `treetoc` inclusions, so 
some warnings during `SIMP_FAST_DOCS=true` are to be expected:

* `SIMP_FAST_DOCS=true` now replaces the dynamic "Known OS
   Compatibility" page with static placeholder text.
  - This skips the 4x release_mappings downloads during every build
     - Now `sphinx-autobuild` doesn't hammer GitHub after every file save
  - Also removes the need to have a simp-core just to build previews
* Updated maximum Sphinx version to 1.8.x (improved error messages)
* `rake clobber` removes content from `docs:html`

----

Finally:
* Added env var `LINKCHECK_IGNORE_LIST` to instruct
  `sphinx-build linkcheck`  to ignore URLs that fail because the CI host 
  has tripped their DDOS protection.

lnemsick-simp commented on Rakefile:
```diff
@@ -444,13 +457,14 @@ which are simply available in the repository.
+      run_streaming_cmd(cmd)
     else
       puts "running web server on http://localhost:#{port}"
       Rake::Task['docs:html'].invoke
```
This task (docs:html) does not run in autobuild_cmd returns nil.  So I don't see any point in invoking it.

lnemsick-simp replied:
Ignore my previous comment,  docs:html doesn't work iv it can't find sphinx-build, not sphinx-autobuild

lnemsick-simp commented on Rakefile:
```diff
@@ -444,13 +457,14 @@ which are simply available in the repository.
     port = args.to_hash.fetch(:port, 5000)
     require 'mkmf'
     autobuild_cmd = find_executable 'sphinx-autobuild'
```
where does sphinx-autobuild come from?  I see sphinx-autogen in my venv created with the requirements.txt, but not sphinx-autobuild

lnemsick-simp replied:
looks like we need a sphinx-autobuild entry in requirements.txt.  Would you please add with the appropriate version ranges you have gotten to work.

op-ct replied:
@lnemsick-simp :ballot_box_with_check:

op-ct commented:
In reponse to some feedback, I tried to get rid of some warning messages when `SIMP_FAST_DOCS=true`, but only with partial success:

* SIMP_FAST_DOCS=true now disables simp-core downloads/mapping
  - replaces "Known OS Compatibility" content with placeholder text_
  - skips simp-core version mapping logic
  - skips github downloads during every build
     - so `sphinx-autobuild` doesn't hammer GitHub after every file save
* `rake clobber` removes html content
* Updated maximum Sphinx version to 1.8.x

After much fussing, I've found that Sphinx is incapable of conditional
`treetoc` inclusions, so the `WARNING: toctree contains reference...`
messages are just to be expected (at least, not without jumping through
much uglier dyanmic content hoops from `conf.py` just for that purpose).

So, I've done a few things to make the situation more obvious:

1. Documented this aspect of `SIMP_FAST_DOCS` in the README and conf.py
2. Added the `security_mapping/index` to the exclusions, so the warnings
   say the document is excluded rather than missing.
3. Upgraded to Sphinx 1.8.x to enable the new warning text

op-ct commented:
Rebased against `simp:master`

op-ct commented:
Rebased against `simp:master`

lnemsick-simp commented on Rakefile:
```diff
@@ -435,12 +450,21 @@ which are simply available in the repository.
+      * SIMP_FAST_DOCS             => (used inside conf.py) defaults to 'true'
+                                       exclude certain long-running sections of
+                                       the documentation (~66% faster)
+      * READTHEDOCS_VERSION        => (used inside docs/conflib/*.py) masquerade as a
```
* Only works when READTHEDOCS='True'.  So, if you want to use that variable, need to mention READTHEDOCS as well.
* IMHO undocumented SIMP_VERSION is more useful in practice.
* Probably should document SIMP_CORE_PATH and SIMP_BRANCH here for completeness. (Cut and paste from README)

lnemsick-simp commented on docs/conflib/get_simp_version.py:
```diff
@@ -45,9 +45,17 @@ def get_simp_version(rootdir=ROOTDIR, simp_github_raw_base=SIMP_GITHUB_RAW_BASE,
         'simp_branch': None
     }
 
+    if os.environ.get('SIMP_VERSION', False):
```
* I get why you did this.  However, I am concerned about the already too many confusing options to get to this information (SIMP_CORE_PATH and SIMP_BRANCH).  It doesn't help that there is code spread over the Rakefile and Python files, the nuance of whether on RTD or not, the nuance of when a `build/rpm_metadata/release` file is generated (processed in the final `else` below),  and a silly option (SIMP_BRANCH)  that doesn't build the version of docs downloaded, only uses the `simp.spec` file from that `simp-core` branch to extract the SIMP version. Can we simplify instead of adding complexity? 
* If you intend to keep this new env variable, please update the unit test for the code changes.

lnemsick-simp commented on docs/conflib/release_mapping.py:
```diff
@@ -157,19 +157,62 @@ def known_os_compatibility_rst(simp_version_dict,
+
+          Also, SIMP_FAST_DOCS is `%s`
+
+      """ % (os.environ.get('SIMP_FAST_DOCS', 'wat'))
```
use this `os.environ.get` in line 182, instead.

lnemsick-simp commented on docs/conf.py:
```diff
@@ -156,8 +156,18 @@ def setup(app):
+# (Sphinx 1.7:)
+#     WARNING: toctree contains reference to nonexisting document u'security_conop/index'
+#
+# (Sphinx 1.8:)
```
This info may need to be documented in Rake task as well.  Unclear if anyone is going to look at the README.

michael-riddle commented:
This change is OBE. Closing.
